tests/std/tests/P0896R4_ranges_alg_is_permutation/test.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <concepts>
+#include <ranges>
+#include <utility>
+//
+#include <range_algorithm_support.hpp>
+
+using namespace std;
+
+struct instantiator {
+    using P = pair<int, int>;
+
+    static constexpr array<P, 8> pairs = {
+        P{0, 42}, P{1, 42}, P{0, 42}, P{1, 42}, P{0, 42}, P{1, 42}, P{0, 42}, P{1, 42}};
+    static constexpr array<int, 8> not_pairs = {0, 0, 0, 0, 1, 1, 1, 1};
+
+    template <class Fwd1, class Fwd2>
+    static constexpr void call() {
+        const Fwd1 range1{pairs};
+        const Fwd2 range2{not_pairs};
+
+        // defaulted predicate and projections
+        assert(ranges::is_permutation(range1, range1));
+        assert(ranges::is_permutation(
+            ranges::begin(range2), ranges::end(range2), ranges::begin(range2), ranges::end(range2)));
+
+        // explicit predicate
+        assert(ranges::is_permutation(range2, range2, ranges::equal_to{}));
+        assert(ranges::is_permutation(ranges::begin(range1), ranges::end(range1), ranges::begin(range1),
+            ranges::end(range1), ranges::equal_to{}));
+
+        // explicit predicate and one projection
+        assert(ranges::is_permutation(range1, range2, ranges::equal_to{}, get_first));
+        assert(ranges::is_permutation(ranges::begin(range1), ranges::end(range1), ranges::begin(range2),
+            ranges::end(range2), ranges::equal_to{}, get_first));
+
+        // explicit predicate and two projections
+        assert(ranges::is_permutation(range1, range2, ranges::equal_to{}, get_first, identity{}));
+        assert(ranges::is_permutation(ranges::begin(range1), ranges::end(range1), ranges::begin(range2),
+            ranges::end(range2), ranges::equal_to{}, get_first, identity{}));
+
+        // negative case
+        assert(!ranges::is_permutation(range1, range2, ranges::equal_to{}, get_second, identity{}));
+        assert(!ranges::is_permutation(ranges::begin(range1), ranges::end(range1), ranges::begin(range2),
+            ranges::end(range2), ranges::equal_to{}, get_second, identity{}));
+    }

```

Comment:
I'd like to see test coverage exercising the metaprogramming for sized/unsized ranges, bidirectional iterators, and the runtime programming for trimming matching prefixes/suffixes, when 1 differing element remains, etc. This is because the algorithm is both nontrivial and a new implementation (instead of a simple copy-paste-modify of the classic algorithm, although they substantially share strategies).

Ideally we should have 100% "code coverage" of all compiletime and runtime control flow, although the cross product isn't necessary.

Reply:
`run_tests` has full coverage for the metaprogramming, but it looks like I'm missing test cases for
* ranges with different lengths, 
* ranges that differ only in their initial element,
* ranges that differ only in their final element,
* element appears more in first range, and
* element appears more in second range
to cover every branch.